File: csharp/0718最长重复子数组.cs
public partial class Solution
{
    public int FindLength(int[] nums1, int[] nums2)
    {
        int m, n, res;
        (res, m, n) = (0, nums1.Length, nums2.Length);

        int[] dp = new int[n];

        // 将m=0时的dp数组初始化
        for (int i = 0; i < n; i++)
        {
            if(nums1[0]==nums2[i])
            {
                dp[i] = 1;
                res = 1;
            }
        }
        
        for (int i = 1; i<m;i++)
        {

            for (int j = n-1; j>0;j--)
            {
                if(nums1[i]==nums2[j])
                {
                    dp[j] = dp[j - 1] + 1;
                    res = Math.Max(res, dp[j]);
                }
                else
                    //注意如果不相等，要重新从0开始计算 
                    dp[j] = 0;
            }
            dp[0] = nums1[i] == nums2[0] ? 1 : 0;
        }
        return res;
    }
}

File: python/main/0718最长重复子数组.py

# dp
def findLength(nums1: list, nums2: list) -> int:
    m=len(nums1)
    n=len(nums2)
    res=0
    dp=[[0]*n for _ in range(m)]

    for i in range(n):
        if(nums1[0]==nums2[i]):
            dp[0][i]=1
            res=1
    for i in range(m):
        if(nums2[0]==nums1[i]):
            dp[i][0]=1
            res=1

    for i in range(1,m):
        for j in range(1,n):
            if(nums1[i]==nums2[j]):
                dp[i][j]=dp[i-1][j-1]+1
                res=max(res,dp[i][j])
    
    return res

print(findLength([1,2,3,2,1],[0]))